ONSdigital/dp-import-cantabular-dimension-options
# use docker-compose to spawn a kafka cluster for component tests

### What

Investigate if we can improve component tests for services using kafka: https://trello.com/c/ONk610ze/5151-investigate-implementing-feature-tests-in-dp-import-cantabular-dimension-options-13

There are 3 options:
- Use `dp-kafka/kafkatest` package, but this is just a mock intended for unit tests, which implements the channels logic, wrapping an moq mock.
- Use `sarama.MockBroker` ( - link - )[https://github.com/Shopify/sarama/blob/master/mockbroker.go], but this is just a TCP Server with some framework to provide mock responses that would be returned by the Kafka broker. It also needs the golang testing package (unit tests), which is not exposed by Armakuni framework. The only advantage of using this over dp-kafka/kafkatest is that we would also test dp-kafka library, having the mock at the network level, instead of the channel communication level.
- Spawn a Kafka cluster with docker-compose, and use real producers / consumers with the right configuration. There could be some argument that the last point would be integration test, but I see this approach as a higher level mock (we would not depend on other services, we just use Kafka producers and consumers to trigger and validate tests)

In this PR I implemented the last approach, without any testing validation yet, as a proof of concept. If we agree, we can start using this approach for component tests of services that involve Kafka.

### How to review

- Make sure code changes make sense
- Run `make test-component` and check that there is no panic.
- Check that the CI runs the component tests successfully: `https://concourse.onsdigital.co.uk/teams/main/pipelines/dp-import-cantabular-dimension-options/jobs/pr-component`

### Who can review

Anyone, specially people that were involved with Armakuni component test works
